portal-kernel/src/com/liferay/portal/kernel/dao/orm/FinderPath.java

```diff
@@ -19,59 +19,113 @@
 import com.liferay.portal.kernel.cache.key.CacheKeyGenerator;
 import com.liferay.portal.kernel.cache.key.CacheKeyGeneratorUtil;
 import com.liferay.portal.kernel.model.BaseModel;
+import com.liferay.portal.kernel.model.CacheModel;
 import com.liferay.portal.kernel.util.HashMapBuilder;
 import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.registry.Registry;
+import com.liferay.registry.RegistryUtil;
+import com.liferay.registry.ServiceRegistrar;
 
 import java.io.Serializable;
 
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
 
 /**
  * @author Brian Wing Shun Chan
  * @author Shuyang Zhou
  */
 public class FinderPath {
 
-	public FinderPath(
+	public static FinderPath create(
 		boolean entityCacheEnabled, boolean finderCacheEnabled,
 		Class<?> resultClass, String cacheName, String methodName,
 		String[] params) {
 
-		this(
+		return create(
 			entityCacheEnabled, finderCacheEnabled, resultClass, cacheName,
-			methodName, params, -1);
+			methodName, params, -1, null, null);
 	}
 
-	public FinderPath(
+	public static FinderPath create(
 		boolean entityCacheEnabled, boolean finderCacheEnabled,
 		Class<?> resultClass, String cacheName, String methodName,
 		String[] params, long columnBitmask) {
 
-		_entityCacheEnabled = entityCacheEnabled;
-		_finderCacheEnabled = finderCacheEnabled;
-		_resultClass = resultClass;
-		_cacheName = cacheName;
-		_columnBitmask = columnBitmask;
+		return create(
+			entityCacheEnabled, finderCacheEnabled, resultClass, cacheName,
+			methodName, params, columnBitmask, null, null);
+	}
 
-		if (BaseModel.class.isAssignableFrom(_resultClass)) {
-			_cacheKeyGeneratorCacheName = _BASE_MODEL_CACHE_KEY_GENERATOR_NAME;
-		}
-		else {
-			_cacheKeyGeneratorCacheName = FinderCache.class.getName();
-		}
+	public static FinderPath create(
+		boolean entityCacheEnabled, boolean finderCacheEnabled,
+		Class<?> resultClass, String cacheName, String methodName,
+		String[] params, long columnBitmask,
+		Function<BaseModel<?>, Object[]> argumentsFunction,
+		Function<BaseModel<?>, Object[]> originalArgumentsFunction) {
 
-		CacheKeyGenerator cacheKeyGenerator =
-			CacheKeyGeneratorUtil.getCacheKeyGenerator(
-				_cacheKeyGeneratorCacheName);
+		FinderPath finderPath = new FinderPath(
+			entityCacheEnabled, finderCacheEnabled, resultClass, cacheName,
+			methodName, params, columnBitmask, argumentsFunction,
+			originalArgumentsFunction);
 
-		if (cacheKeyGenerator.isCallingGetCacheKeyThreadSafe()) {
-			_cacheKeyGenerator = cacheKeyGenerator;
-		}
-		else {
-			_cacheKeyGenerator = null;
+		ServiceRegistrar serviceRegistrar = _serviceRegistrars.computeIfAbsent(
```

Comment:
Why ServiceRegistrar here? Shouldn't we just hold them as a map of ServiceRegistration?

Reply:
@shuyangzhou I just want it to be easy to destroy.

Reply:
ServiceRegistration is as simple as this, and cheaper, switch to ServiceRegistration directly, no need of this wrapper.

Reply:
@shuyangzhou If I use ServiceRegistration, one option is have one layer of map, which means I have to iterate every entry when removing, another option is to have two layers of map which may make it look more complicate. So why you do not like ServiceRegistrar? for performance or for readability?

Reply:
Right now you are only unregistering the FinderPath service when the cache goes away, right? That's not enough for later. Later when DSL query comes in, from module, we might need to register/unregister FinderPath just by itself, without taking down the cache or siblings FinderPath.
This also brings the problem of those factory methods in FinderPath. The creation and registeration of FinderPath needs to be separated, otherwise, DSL query users won't be able to get access to the ServiceRegisteration for those FinderPath. The point is, the owner of the FinderPath should control the service lifespan.

Reply:
Right. My plan is to extract more finder path generation logic out from persistent impl, I think I can register a set of path at the same time, then return the service registrations back, so that the client will control the lifecycle of finder path service themselves, meanwhile we won't have the long list of finder path field in ***PersistentImpl.

Reply:
For FinderPath usages inside PersistenceImpl, you can do that. But it needs to be done inside PersistenceImpl, not FinderPath itself. FinderPath should not handle its own registeration. It needs to leave the option open to users.

Reply:
If we did not force the registration, what if people forgot to register? we should not allow a finder path used without registration. As long as I give back the service registration to user, the user will have full control, right?

Reply:
yeah, that's fine, as long as the owner controls the registration.